### ISSUE TITLE
EPIC: Support iterative reading of sparse6 and incremental sparse6 formats

### ISSUE THREAD

**Issue by john-boyer-phd** [graph-algorithms/edge-addition-planarity-suite#38]
Nauty supports output of a ["sparse6" and incremental "sparse6" file formats](https://users.cecs.anu.edu.au/~bdm/data/formats.txt) to contain all graphs generated according to its parameterization. Subject to some experimentation, it is likely that the incremental "sparse6" format could significantly improve performance of algorithm testing, i.e., performance and output from command lines of the form `planarity -t [-q] C I O`. 

- [ ] Begin by using Nauty to generate all simple undirected graphs of order 10 and each of the possible numbers of edges in separate files. Do this for both the "graph6" and the incremental "spase6" formats so that we can get a sense of when "graph6" is less versus more efficient in terms of storage than incremental "sparse6". It may be necessary to also do this for order 9 and 11 to get a better sense.

The incremental "sparse6" format is expected to helps in the following ways:
- More of the files generated may be able to fit entirely into memory such that read iteration of graph reading occurs over an in-memory string rather than against a file. 
- Rather than reinitializing and fully rebuilding each graph to be tested with an algorithm, the incremental format offers the possibility of  simply adding an edge or edges to a previously given graph.

If the incremental "sparse6" format looks promising enough, then further work will be done to define the work items in this epic.

For transformation of the first graph in a file to the "sparse6" format, the letter 's' is used, i.e., `planarity -t [-q] -ts I O`. Since only one graph is transformed, the incremental format is not applicable as an output format.

**Comment by Deln0r** [comment]
I would like to take this EPIC, starting with the measurement checkbox: orders 9 through 11 generated in graph6 and incremental sparse6, storage and read-throughput compared, results posted here before any code. The reader itself I would build the way I build wire-format ports elsewhere: a fixture set pinned against nauty's own tools as the reference, exercised in both directions, then the iterator implemented to the same API shape as the existing g6 one. If you or Wanda have design expectations beyond what is written in the issue, tell me before I start.

**Comment by Deln0r** [comment]
Measurements for the first checkbox, before any code. Method: `geng -q` in its canonical order, conversions with `copyg -s` and `copyg -i`, read cost measured as `copyg -g` into `/dev/null`, so the reader is nauty's own. Apple M3, single runs; sizes are exact, times are indicative.

| order | graphs | graph6 | sparse6 | incremental s6 | is6/g6 | read g6 | read is6 |
|---|---|---|---|---|---|---|---|
| 9 | 274,668 | 2.20 MB | 5.29 MB | 1.49 MB | 0.68 | - | - |
| 10 | 12,005,168 | 120.1 MB | 276.1 MB | 65.3 MB | 0.54 | 2.17 s | 1.98 s |
| 11 | 1,018,997,864 | 12.23 GB | (skipped, see below) | 5.56 GB | 0.45 | 216 s | 171 s |

Per edge count at order 10, is6/g6 stays between 0.54 and 0.63 across the whole density range, with the single complete graph as the one outlier (format overhead on a one-graph file). So the win is not a sparse-graphs-only effect: it comes from geng's generation order, where consecutive graphs differ little and the incremental encoding pays for exactly that.

Two conclusions I would draw for this EPIC. Plain sparse6 is not worth a reader for enumeration streams: it more than doubles graph6 on these files, and I skipped the 28 GB order-11 conversion on that basis. Incremental sparse6 is the target: the order-11 inputs halve and read faster. Next, per my plan above: a fixture set pinned against copyg in both directions, then the read iterator to the g6 API shape.

**Comment by Deln0r** [comment]
Reader is up as #329. Same API shape as the g6 iterator, decoded straight from the stream (no line buffer, since sparse6 lines have no fixed length), and incremental lines toggle edges on the previous graph instead of rebuilding it. Pinned against `copyg` in both directions on every graph of orders 2 to 8 plus the order-10 set; reading the incremental order-10 file takes 1.6 s against 3.0 s for graph6 on the same 12,005,168 graphs. Next is `planarity -t` over sparse6 input, where the incremental path should show the real gain because the rebuild per graph goes away.

**Comment by john-boyer-phd** [comment]
@Deln0r Hi Ian, It seems likely that sparse6 writer support will require an additional API function to help operate the incremental case. Without an extra function, every time one calls `s6_WriteGraph()`, one would think that it would write a full graph on a `:` line.

It would be nice to have an extra public function with a signature like `int s6_StoreGraphChange(theS6WriteIterator, int e, int u, int v)`. If `e` is non-NIL, then this call should store an edge deletion record into a growing array of changes. If `u` and `v` are non-NIL, then store an edge addition record into the growing array of changes. Other possibilities should produce invalid parameters `NOTOK` error. 

Then, when `s6_WriteGraph()`, the following things would happen:

1. If there is a non-empty array of changes, then write the changes as a `;` line into the output.
2. If there is a non-empty array of changes, then apply the changes to the graph within the write iterator.
3. If there is a non-empty array of changes, then set the array of changes back to a zero size, since they have been processed.
4. On the other hand, if the array of changes is empty when `s6_WriteGraph()` is called, then we write the full graph in the write iterator in a `:` line because we must assume that the caller wants the full graph written and has made whatever changes they want beforehand.

An additional reason that I'm suggesting this approach is that it is aligned with a future improvement of the reader that I have reason to need. I would eventually like to have an `int s6_RetrieveGraphChange(theS6ReadIterator, int *e, int *u, int *v)` that could tell me what if any incremental changes are coming. If the next line is a `:`, then all three pointers are assigned `NIL` as the way to detect that. Also, same three `NIL` values when we reach the end of `;` line. These invocations should grow an array of what has been read, and when we call `s6_ReadGraph()`, then any changes in the array are applied to the graph currently in the read iterator, and then the change list is cleared. If the caller `s6_ReadGraph()` without calling `s6_RetrieveGraphChange()`, then the change list is empty, and the next graph is read whether it is a `:` line or a `;` line. The main idea is that we will eventually need lookahead-style reading for the incremental case, which is analogous to developing a list of changes for incremental writing.

This sound OK?

**Comment by Deln0r** [comment]
Sounds right. The reader already applies a `;` line as toggles, so the writer mirrors it. Contract points:

1. StoreGraphChange validates at call time: absent edge on delete, present edge on add, loop, or a repeated pair return NOTOK.
2. NOTOK before the first `:` line is written.
3. A batch is relative to the last written graph, so direct edits need a full `:` write first. The writer cannot detect them, so the header must say it.
4. Store endpoints, not edge indices, so one deletion cannot stale another.

Lookahead: ReadGraph must finish a partly retrieved line, and an empty `;` line read ahead still counts as a graph.

`-t` over sparse6 input first, then the writer.

**Comment by john-boyer-phd** [comment]
**Regarding point 1:**

The way you wrote this caused me to understand that we either don't have quite the same idea for the function or that you may be pointing out additional error cases.

The caller owns the graph in the write iterator and is reasonably expected to know what is in it.

My hope is that the caller would call `s6_StoreGraphChange(theS6WriteIterator, e, NIL, NIL)` to indicate that they want edge `e` deleted from the iterator's graph before writing. So, if the edge value given fails to be within the lower and upper bounds for in-use edges, or if it fails `gp_EdgeInUse()`, then a `NOTOK` should occur and no change should be stored. Since the edge does exist, writing it to the `;` line will signal to turn it off, and it will be appropriate to call `gp_DeleteEdge()` to get rid of it from the iterator's graph when processing the change list during the write graph operation.

I also hope that the caller would use `s6_StoreGraphChange(theS6WriteIterator, NIL, u, v)` to mean that they would like to have the edge `(u, v)` added to the graph. Presently, it should be `NOTOK` to have `u == v` (deny loops), and `NOTOK` for `u` or `v` to violate the lower and upper bound for vertex indexes, and `NOTOK` to have `gp_FindEdge(theGraph, u, v)` return a non-`NIL` value (deny parallel edges).  Support for both loops and parallel edges is definitely coming in this release, but later, so we can presently deny them and later adapt the sparse6 code as part of providing that multigraph support.

**Regarding point 2:** 

Yes, good catch. Confirmed: For write iteration, the caller is assumed to have populated the graph before calling `WriteGraph` for the first time, and that whole graph needs to be written before the caller starts to store changes to it. 

**Regarding point 3:** 

The first sentence seems to be the same as your point 2, so agreed that "A batch is relative to the last written graph, so direct edits need a full `:` write first." 

I didn't understand the second sentence: "The writer cannot detect them, so the header must say it." Could you say more about this?

I think that the writer can detect whether it has written at least one graph just using a boolean state variable. Once that criterion is met for the first time, then the caller should be able to call `s6_StoreGraphChange()` zero or more times before calling `s6_WriteGraph()` again. If the caller has stored at least one change before calling `s6_WriteGraph()`, then the changes are applied and then the changes are written as a `;` line. On the other hand, if there are zero changes stored when `s6_WriteGraph()` is called, then it writes the whole graph as a `:` line. Does that sound OK?

**Regarding point 4:**

It would be really handy to add a function to `graph.c` whose signature would be added to `graph.h` in between `gp_EnsureEdgeCapacity()` and `gp_ResetGraphStorage()`. The function signature would be `int gp_CompactEdgeStorage(graphP theGraph)`. The implementation would systematically pop edge holes and fill them until all edge holes are filled and the `edgeHoles` stack is empty. Should return `OK` unless the caller passes a `NULL` graph.

Once it exists, this function could be called immediately before the end of `s6_ReadGraph()` to get rid of edge holes as a batch operation, rather than getting rid of them incrementally as you currently do. For the writer, the group of stored graph changes could all be processed, and then `gp_CompactEdgeStorage()` could be called after all stored changes have been processed. It is not necessary to do this for the writer, but it would be clean and also symmetric with the read case. Either way, for both the reader and writer, the batch approach to eliminating edge holes would eliminate the valid concern you raised in point 4 about an edge deletion making stale the edge identifiers of subsequent deletions. 

The reason I am mentioning it is that the use case I have in mind for the reader is that the caller will need to run a test to check the nature of edges to be deleted. For simple graph processing, it is only that it is more efficient to receive an indication of the edge that will be deleted, rather than receiving the endpoints and then finding the edge. However, multigraph support is coming very soon, in the5.2 release. For a multigraph, the caller would need to know which instance of an edge between the same endpoints is the one that the reader will delete. The reader will not know from the file content which instance to delete, so it will select one, but the reader is behind the API wall, and the caller should not make an assumption about which instance the reader will select, which is solved by the reader indicating each edge it will select for deletion in the `s6_RetrieveGraphChange()` calls.

This also means that `s6_RetrieveGraphChange()` will be analogous to `s6_StoreGraphChange()`. If a value for `e` is returned via the output parameter `int *e` then `s6_RetrieveGraphChange()` is indicating that the read iterator's graph has an edge `e` that the current `;` line is togging to be deleted. Similarly, if values are instead returned for `u` and `v`, then `s6_RetrieveGraphChange()` is indicating that it will be adding a new edge to the read iterator's graph when `s6_ReadGraph()` is called.

Does all of this sound OK?

**Comment by Deln0r** [comment]
All of it sounds OK. On the second sentence of point 3: the boolean covers the first write. The case I meant is a caller that edits the graph directly (gp_AddEdge, gp_DeleteEdge) after that write and then stores a batch. The `;` line would be relative to the graph as last written, and the writer cannot notice, so the header should say that between writes the graph changes only through StoreGraphChange, or the next write is a full one.

One addition to point 1: validation looks at the unmodified graph, so a pair stored twice passes both times; the batch itself needs the repeat check.

gp_CompactEdgeStorage() after ReadGraph and after the batch: yes, cleaner than my per-deletion fill.

**Comment by john-boyer-phd** [comment]
Two excellent points Ian @Deln0r 

Regarding the issue of the caller editing the graph directly while also setting up a batch of changes, it is an "unhappy path" through the API, so it would be reasonable to return `NOTOK` if we detect a change to the graph. The graph can be modified by many methods such as `gp_InsertEdge()`and `gp_DeleteEdge()` but also `gp_CopyGraph()`, `gp_CopyAdjacencyLists()`, `gp_ResetGraphStorage()`, and probably more. What do you think?

One way to have a catch-all for detecting if the graph structure was modified by any methods that exist now or that might exist in the future is to take an internal copy of the graph at the time that a first change is stored (using `gp_DupGraph()`). Then, at the point when you're about to resolve a change list, compare the duplicated graph with the write iterator's graph edge for edge in a subroutine. This could be done by running the adjacency lists of each vertex or by running through the in-use edges in the edge array. Any discrepancies means return `NOTOK` rather than applying the change list. How does this sound?

Also, good point about the change list possibly having duplicate add edges. We are increasing parallel edge support in this release, and `gp_InsertEdge()` already does allow it, but it's probably best for you to detect and forbid it now, and then we can later modify that to say that it will set the parallel edges flag instead.

**Comment by Deln0r** [comment]
The duplicate-and-compare is safe but costs O(E) time and a second graph per batch, and the point of `;` lines is many small batches on large graphs, so it would take back most of the gain. I would rather keep the writer O(batch): a modification counter in the graph, bumped by every mutator (gp_InsertEdge, gp_DeleteEdge, gp_CopyGraph, gp_ResetGraphStorage, and any future one), snapshotted at each write, compared before a batch is resolved; mismatch is NOTOK. The counter is a convention new mutators have to follow, so the full comparison can stay as a DEBUG-build check that catches a mutator that forgot. Duplicate adds forbidden now, agreed.

**Comment by john-boyer-phd** [comment]
Ah, you're right. I usually don't think as much about cost when disk writing anyway, but O(E) for a `;` line is a bit far.

May I suggest putting the modification counter in `graphPrivateDataStruct` (and its getter/setter functions in `graph.private.h`)? Using `graphPrivateDataStruct` is important so we don't break backwards compatibility. Putting the getters and setters in the package private header means sparse6 can use them, but graph application developers cannot have these right now until/unless we later decide this is what we want them to have.

There are a few very low-level macros that can change structure that might technically count as modifications but which I'd prefer not to make any more costly by adding a pointer access modification increment. Pop culture reference: _First rule of planarity club is do not affect the speed of planarity, bub_. Incrementing the modifier in the random graph generators would be good though.

Thanks, Ian!

**Comment by john-boyer-phd** [comment]
The following seem safe for flagging modification: 
 * copy graph, 
 * copy adjacency lists, 
 * create random graph, 
 * create random graph ex, 
 * insert edge (no need to do add, dynamic add, and dynamic insert, since they call insert),
 * delete edge, 
 * hide/contract/identify/restore edge/vertex/vertices (7 functions),
 * sort vertices

Other functions and macros seem unrelated to this, except for some setter macros related to first/last/next/prev/byLink/adjacent edge. These are available but if we're worried about them, then we'd have to be worried about someone directly accessing the data fields, too, and that is actually where the compare graphs idea came from. So, it seems better to conclude that someone is on an unhappy path that is just not supported.

**Comment by Deln0r** [comment]
Placement sounds right, and private accessors are enough for sparse6.

Two refinements after reading master. The hide and restore family funnels through the public wrappers: hide vertex, contract edge, identify vertices and restore vertices all reach storage through gp_HideEdge or gp_RestoreEdge, so increments in those two, plus gp_IdentifyVertices and gp_RestoreVertex where links are also rewritten directly (graph.c:2659 and 2774), cover all seven. And graphDFSUtils.private.h already includes graph.private.h, so sort vertices needs no new include.

Insert covers gp_AddEdge, both dynamic variants and both random generators, which build their edges through gp_AddEdge. Worth adding gp_ResetGraphStorage in its wrapper, and gp_TransposeDirectedGraph with gp_ClearEdgeDirectionFlags, even though sparse6 refuses digraphs.

**Comment by Deln0r** [comment]
Step three is up as #341: `-t` takes sparse6 and incremental sparse6 input, with the format read from the content of the file rather than its name.

The measurement, on the same 12,005,168 graphs of order 10 as before, three interleaved rounds each. `-t -p` takes the same user time either way, 22.9 to 23.3 s from `.g6` against 23.2 to 23.5 s from `.inc.s6`, with identical counts. The incremental file is 65.3 MB against 120.1 MB and reads faster on its own, 1.8 s against 2.1 s through `copyg -g`, but reading is a few percent of the run: `gp_CopyGraph()` and `gp_Embed()` dominate. So on this workload the format buys disk and read time, not embedding time, which is worth having written down before the writer goes in.

Building it also turned up a one-byte write past the graph6 reader's line buffer, in the pushback path that `gp_Read()` has been using all along; that is #339, and the writer PR sits on top of its fix.

Writer next, to the API we settled above, with `gp_CompactEdgeStorage()`.

**Comment by Deln0r** [comment]
Step four is up as #344: the writer, to the API above. `s6_StoreGraphChange()` validates at store time, `s6_WriteGraph()` writes a `;` line for a pending batch and a `:` line otherwise, and the modification counter, in the private data with package private accessors, is what refuses a batch after a direct edit. `gp_CompactEdgeStorage()` went in as agreed, and the reader uses it too.

The bytes match `copyg -s` and `copyg -i` on all 12,293,433 graphs of orders 2 to 10 and on random graphs to order 1000. The lookahead stays for a later PR, at the `gp_` level as you suggested on #341.

**Comment by john-boyer-phd** [comment]
Hi Ian (@Deln0r ), My apologies for one confusion I created in a note on #341, which relates to this comment in your PR #344 

> Not in this PR: the reader lookahead s6_RetrieveGraphChange(), which per your note on https://github.com/graph-algorithms/edge-addition-planarity-suite/pull/341 belongs at the gp_ level, with every format but sparse6 reporting no increment.

I should clarify that I didn't mean for there to be no `s6_RetrieveGraphChange()`. Instead, what I meant was that afuture PR which offers the incremental reading function to the API user should _also_ have a `gp_RetrieveGraphChange()`. It can just call `s6_RetrieveGraphChange()` if the file is an s6 and otherwise just give the "no incremental change" return result for the other file types. So, if someone is using the `s6` API, they can stick with it to get this feature, and if they are using the `gp` API, then they can stick with that, instead of having to drop down into the `s6` layer. Also, having the `gp` layer function makes it possible for other formats to offer the incremental feature, too, as has apparently been occasionally done with both GraphML and GML.

**Comment by Deln0r** [comment]
Thanks, John, and the ambiguity was in my line, not your note. So the step after #344 is one PR with both: `s6_RetrieveGraphChange()` on the sparse6 read iterator, and `gp_RetrieveGraphChange()` on the gp_ reader, which calls it for sparse6 and returns "no incremental change" for every other type. The contract stays as settled above: the edge id for deletions, three NILs when the next line is a `:` line or the `;` list is done, and `s6_ReadGraph()` finishing a line the lookahead has started.

**Comment by john-boyer-phd** [comment]
Just a quick note to let you know that in case you saw the #344 PR review with only three items, I had to edit it to add a fourth item because I realized a problem that could occur with `gp_CompactEdgeStorage()` as implemented if it were used outside of the use case you currently have.
